modules/packages/RangeChunk.chpl

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright 2004-2016 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ * 
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * 
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+module RangeChunk {
+  enum RemElems {
+    Thru,
+    Pack,
+    Mod
+  }
+  use RemElems;
+  use BoundedRangeType;
+
+  /*
+   * Range-based interface, iterator and query.
+   * idx in query is zero based.
+   */
+
+  iter chunks(
+    r: range(?RT, bounded, ?S),
+    numChunks: integral,
+    remPol: RemElems = Thru): range(RT, bounded, S)
+  {
+    for (startOrder, endOrder) in chunksOrder(r, numChunks, remPol) {
+      const start = r.orderToIndex(startOrder);
+      const end = r.orderToIndex(endOrder);
+      yield if S
+        then start..end by r.stride
+        else start..end;
+    }
+  }
+
+  proc chunk(
+    r: range(?RT, bounded, ?S),
+    numChunks: integral,
+    idx: integral,
+    remPol: RemElems = Thru): range(RT, bounded, S)
+  {
+    const (startOrder, endOrder) = chunkOrder(r, numChunks, idx, remPol);
+    const start = r.orderToIndex(startOrder);
+    const end = r.orderToIndex(endOrder);
+    return if S
+      then start..end by r.stride
+      else start..end;
+  }
+
+  /*
+   * Order-based interface, iterator and query.
+   * Orders are zero-based "indices" into a range. 
+   * idx in query is zero-based.
+   */
+
+  iter chunksOrder(
+    r: range(?RT, bounded, ?),
+    numChunks: integral,
+    remPol: RemElems = Thru): 2*RT
+  {
+    if r.length == 0 || numChunks <= 0 then
+      return;
```

Comment:
I wanted to convey an empty iterator, because there was a test case calling this with an empty range. I wasn't sure of the best way to do it, so I went with this ``void`` return.

Is there a better way to address this?